// rumble-charts/lib/DropShadow.js
'use strict';

var React = require('react');

var DropShadow = React.createClass({

    displayName: 'DropShadow',

    propTypes: {
        id: React.PropTypes.string.isRequired,
        dx: React.PropTypes.number,
        dy: React.PropTypes.number,
        blurDeviation: React.PropTypes.number,
        blurIn: React.PropTypes.oneOf(['SourceAlpha', 'SourceGraphic', 'BackgroundImage', 'BackgroundAlpha', 'FillPaint', 'StrokePaint'])
    },

    // init

    getDefaultProps: function getDefaultProps() {
        return {
            dx: 1,
            dy: 1,
            blurDeviation: 4,
            blurIn: 'SourceAlpha'
        };
    },


    // render

    render: function render() {
        var props = this.props;

        return React.createElement(
            'filter',
            { id: props.id },
            React.createElement('feGaussianBlur', { 'in': props.blurIn, stdDeviation: props.blurDeviation }),
            React.createElement('feOffset', { dx: props.dx, dy: props.dy }),
            React.createElement(
                'feMerge',
                null,
                React.createElement('feMergeNode', null),
                React.createElement('feMergeNode', { 'in': 'SourceGraphic' })
            )
        );
    }
});

module.exports = DropShadow;
